File: backend/endpoints/PostEndpoint.cs
using backend.Service;

namespace backend.Endpoints;

public static class PostEndpoint
{
  public static IServiceCollection UsePostApi(this IServiceCollection service)
  {
    service.AddScoped<IPostService, PostService>();
    return service;
  }

  public static WebApplication MapPostEndpoint(this WebApplication app)
  {
    app.MapGet("/api/posts", (IPostService postService) => postService.GetNewestPostAsync());

    app.MapGet("/api/posts/{id}", (IPostService postService, int postId) => postService.GetPostById(postId));

    app.MapPost("/api/posts", (IPostService postService, string content) => postService.CreatePostAsync(content));

    app.MapPut("/api/posts/{id}", (IPostService postService, int id, string content) => postService.UpdatePostAsync(id, content));

    app.MapDelete("/api/posts/{id}", (IPostService postService, int id) => postService.DeletePostAsync(id));

    return app;
  }
}

File: backend/endpoints/PostEndpoint.fs
module Backend.Endpoints.PostEndpoint

open System
open Microsoft.AspNetCore.Builder
open Backend.Services.PostService
open Microsoft.Extensions.DependencyInjection
open Microsoft.AspNetCore.Http

let usePostEndpoint (app: WebApplication): WebApplication =
  
  app.MapGet("/api/posts", Func<HttpContext, string>(fun httpContext ->
    let postService = httpContext.RequestServices.GetRequiredService<PostService>()
    postService.getLatestPostHTMLContent()
  )) |> ignore

  app.MapGet("/api/posts/{id}", Func<HttpContext, int, string>(fun httpContext id ->
    let postService = httpContext.RequestServices.GetRequiredService<PostService>()
    postService.getPreviousPostHTMLContent(id)
  )) |> ignore

  app.MapGet("/api/posts/titles", Func<HttpContext, string>(fun httpContext ->
    let postService = httpContext.RequestServices.GetRequiredService<PostService>()
    postService.getAllPostTitlesAsHTMLUl()
  )) |> ignore


  app
